Replace yarn 1 with pnpm

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Replace yarn 1 with pnpm.

A few side effects:

* Update a bunch of indirect dependencies.
* Add some missing dependencies.
* Move the js-test-runner into its own "project", like the Jetpack CLI, so we can remove its deps from the monorepo-root package.json.
* Deprecated a bunch of scripts in the monorepo-root package.json.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p9dueE-2R4-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

See [docs/yarn-upgrade.md](https://github.com/Automattic/jetpack/blob/feature/replace-yarn-1-with-pnpm/docs/yarn-upgrade.md) before you start, you'll probably want to do some cleanup before checking this out.

* Try everything, see if it still works.

anomiex committed:
Switch to pnpm: Make `pnpm install` (and precommit hooks) work

* Convert yarn's "resolutions" to pnpm's "overrides". Clean up several
  that were obsolete.
* Convert yarn.lock to pnpm-lock.yaml. This was a little roundabout: First
  I used `npm install --legacy-peer-deps --package-lock-only` to let npm 7
  convert yarn.lock to package-lock.json, then `pnpm import` to convert
  that to pnpm-lock.yaml.
* Set up a workspace. This lets intra-monorepo dependencies work more
  sensibly.
* Update tools/prettier and tools/git-hooks/pre-commit-hook.js to run
  things via pnpm/pnpx rather than yarn, so `git commit` works.

Of note is that the `pnpm import` seems to have deduplicated/updated
some indirect dependencies, and then the switch to a workspace seems to
have done a bit more of that. Hopefully that doesn't break anything.

anomiex committed:
Switch to pnpm: Monorepo package.json and composer.json

We have a lot of scripts defined at the monorepo level that seem to
exist for the purpose of building the Jetpack plugin. That should really
be done by the Jetpack CLI instead, so let's deprecate them all.

anomiex committed:
Switch to pnpm: Disable use of yarn

Since everyone is probably in the habit of running "yarn", set a bogus
version in package.json's .engines.yarn so it'll give an error instead
of trying to run.

anomiex committed:
Switch to pnpm: Tools

Of note is that `pnpm link --global` looks in `$PATH` to decide where to
link the executable, versus yarn and npm which just add to a particular
directory. And when `pnpm cli-link` is used to run `jetpack cli link`,
it will have prepended several node_modules/.bin directories (and a
node-gyp-bin dir) to the path, so `jetpack cli link` needs to strip
those off before in turn calling `pnpm link --global`.

I wonder whether all the places where something says "run `jetpack foo`"
should actually be "run `pnpx jetpack foo`" for people who haven't done
the cli-link thing.

anomiex committed:
Switch to pnpm: GitHub actions

This uses the pnpm/action-setup action to install pnpm.

I wish GH would provide a way to reuse a sequence of steps that use
other actions, instead of us having to repeat basically the same 59-line
block of stuff everywhere to set up php, node, and pnpm.

anomiex committed:
Switch to pnpm: Jetpack plugin

anomiex committed:
Switch to pnpm: All other projects

anomiex committed:
Add a dummy pnpm-workspace.yaml in projects/plugins/jetpack/tests/e2e

Otherwise `pnpm install` in that directory gets confused and operates on
the workspace instead of the non-workspace package.json in that
directory.

anomiex committed:
Update lockfiles

anomiex committed:
pnpm-lock: Use @babel/compat-data 7.8.6 with @babel/preset-env 7.8.4

The old yarn.lock had this, and it breaks if a later version is used.

anomiex committed:
Jetpack: Cleanup js deps

* @babel/plugin-transform-runtime requires @babel/runtime, and tell
  webpack where to look for it.
* Several other babel things seem like dev deps rather than runtime
  deps.
* Calendly block depends on `query-string`.
* Several blocks depend on `url`.
* Stuff depends on `@wordpress/blocks` and `@wordpress/hooks`.

anomiex committed:
Clean up root deps by moving js-test-runner to a workspaced project

I wanted to do this back in #19140, but at the time yarn bugs prevented
it.

anomiex committed:
Downgrade @wordpress/blocks and @wordpress/rich-text

Newer versions make Jetpack's build fail.

anomiex committed:
Jetpack: Add missing default node_modules path in webpack.config.search.js

anomiex committed:
Jetpack: Use require.resolve in webpack.config.search.js aliases

I wish I understood why this works.

anomiex committed:
Restore some locks to make test-extensions pass

@wordpress/block-editor@^3.7.7 -> 3.7.7
@wordpress/element@^2.11.0 -> 2.18.0
@wordpress/element@^2.18.0 -> 2.18.0
@wordpress/rich-text@^3.12.2 -> 3.12.2
react@^16.13.1 -> 16.13.1
react@^16.8.3 -> 16.13.1

anomiex committed:
cli: Add missing dep on js-yaml

anomiex committed:
Add `--no-prod` to `pnpm install` calls in package.jsons

anomiex committed:
Fix yarn calls added after rebase

anomiex committed:
Lock parse-asn1 to 5.1.5

5.1.6 upgrades its dep on asn1.js to a version that needs transpilation.

anomiex committed:
Jetpack: Fix non-exclusion regex in webpack.config.search.js

Paths now have two `/node_modules/`, one followed by `.pnpm/` and the
other by the package name.

anomiex committed:
Use published version of action-required-review

anomiex committed:
Add chalk to monorepo root

For tools/check-renovate-ignore-list.js

anomiex committed:
required-review.yml: Still need checkout for definition file

anomiex committed:
js-packages/connection: Add missing lodash dep

anomiex committed:
Jetpack: Update composer.lock

anomiex committed:
js-test-runner: Fix lint

anomiex committed:
Jetpack e2e: wp-env is a bin, not a script

anomiex committed:
Jetpack e2e: pm2 is too

anomiex committed:
Jetpack e2e: Remove yarn workaround from execWpCommand

It was yarn, not `wp-env run`, that printed the command run. Since pnpx
doesn't do that, remove the code trying to remove it.

anomiex committed:
Add change entries

anomiex commented:
* Tests / JS tests is currently failing due to https://github.com/pnpm/pnpm/issues/1768, which should be fixed by https://github.com/pnpm/cmd-shim/pull/36. That PR is already approved upstream, hopefully they merge it and do a release sooner rather than later.
* Tests / Code coverage looks to be failing for the same reason.
* E2E Tests / E2E connection tests is failing on other PRs with the same error, so probably nothing here.

anomiex committed:
Jetpack: Better fix for babel locating @babel/runtime

It turns out @babel/plugin-transform-runtime already has an option for this. We just have to hack up what gets returned from calypso-build to set it.

anomiex committed:
Replace "yarn" with "pnpm" in docs

As far as what's listed in the docs, it turns out pnpm is mostly a
drop-in replacement. The main differences are:

* You have to use `pnpm` for scripts versus `pnpx` for bins, while
  `yarn` handles both.
* When running a script via `pnpm`, you need a `--` before any options
  that are supposed to be passed to the script. `yarn` does not require
  this.

I also saw a lot of places where the docs are outdated, e.g.
"Development workflow" in docs/development-environment.md seems to have
never been updated for the changed monorepo layout. Fixing that seems
out of scope for this PR.

anomiex committed:
Add change entries for doc updates

anomiex committed:
e2e-tests: Install pnpm in "Slack notification" job

kraftbj commented:
I'm aiming to review today. Marking as blocked since we'll need communications in place before merge to alert folks of required tooling change.

jeherve commented:
At this point of the release cycle, would it make sense to wait until after the code freeze / branch off `master` to merge this, to avoid any potential issues that may come up? Merging this early in the release cycle would leave us a full cycle to iron things out.

anomiex commented:
I think the necessary communications will wind up requiring we hold off until the next release cycle anyway. I'm hoping we don't wind up having to wait longer, actually, for things like p1621520158124600-slack-CBG1CP4EN.

kraftbj commented on docs/development-environment-cli-support.md:
```diff
@@ -12,15 +12,13 @@ tools/check-development-environment.sh
-If `jetpack` is still not available, you can run `yarn jetpack` from the repo root only.
-
-`jetpack`, when installed, will work from anywhere within the repo.
+You can also use `pnpx --no jetpack` from (almost) anywhere in the monorepo. If you commonly use multiple checkouts, you might add this as a shell alias or a stub script in your path so as to always run the current checkout's instance.
```
Nice add.

kraftbj commented on docs/development-environment.md:
````diff
@@ -167,34 +167,34 @@ There are three types of builds:
 
 ```sh
-yarn distclean
+pnpm distclean
````
This is fine for now I think. It mentions it is deprecated for me, but we can wait until the CLI clean command is ready before updating the docs.

anomiex replied:
Yeah, as I noted in the commit message on eb625dfdc7 updating all the outdated docs seems out of scope for this PR.

kraftbj commented on package.json:
```diff
@@ -85,34 +85,15 @@
-		"**/terser-webpack-plugin/serialize-javascript": "^5.0.0"
-	},
 	"dependencies": {
+		"chalk": "4.1.1",
```
Only a couple of deps we hadn't declared. That's not too bad, methinks.

anomiex replied:
I didn't try running all the tools manually, but we don't have too many that are written in js.

kraftbj commented on projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -174,7 +174,7 @@ function page_render() {
 			echo '<p>';
 			esc_html_e( 'Error fetching static.html. Try running: ', 'jetpack' );
-			echo '<code>yarn distclean && yarn build</code>';
+			echo '<code>pnpm run distclean && pnpm run build</code>';
```
While we're here, should update this to the cli command.

```suggestion
			echo '<code>pnpm run distclean && pnpx jetpack build plugins/jetpack</code>';
```

kraftbj reviewed (dismissed):
Still want to do some more testing and didn't quite get through all of it yet, but here's a few minor things so far.

kraftbj commented:
Upon switching to the branch:
![2021-05-20 at 10 14 AM](https://user-images.githubusercontent.com/88897/119004526-4327b800-b954-11eb-90d6-7ae1bafc9296.png)

Following the directions to delete `node_modules` resolves it (with the usual error about husky-run when the repo isn't installed), but should be ready to explain to folks what this means post-merge. I'm looking at making a script to help with the transition; I'll PR that separately when I get it ready-enough.

anomiex committed:
Apply suggestions from code review

Co-authored-by: Brandon Kraft <public@brandonkraft.com>

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex committed:
Update pnpm version

Version 6.5.0 contains the fix for the NODE_PATH bug.

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex committed:
Fix more deps

simple-git doesn't need to be in the root, tools/cli is what needs it.

Jetpack needs semver, so add it there instead of using the root dep.

anomiex committed:
Add upgrade doc

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex committed:
doc tweaks

anomiex committed:
Let pnpm update some indirect deps

Let's see if this actually breaks anything before I spend time trying to
make it not update them.

anomiex committed:
Yep, that broke something

@wordpress/rich-text 3.12.2 depends on @wordpress/data ^4.14.2, but
breaks with versions after 4.23.1. Add an override for that.

jeherve commented on package.json:
```diff
@@ -13,65 +13,65 @@
+		"docker:stop": "pnpm run docker:compose stop",
+		"docker:tail": "pnpm run docker:compose exec wordpress bash -c \"/var/scripts/tail.sh\"",
+		"docker:uninstall": "pnpm run docker:compose -- exec wordpress bash -c \"/var/scripts/uninstall.sh\"",
+		"docker:up": "pnpm run docker:compose up",
```
Should it be `pnpm run docker:compose -- up` here?

```suggestion
		"docker:up": "pnpm run docker:compose -- up",
```

I used to run `yarn docker:up -d` to have my container running in the background, but this now only works if I do `pnpm run docker:compose -- up -d`.

anomiex replied:
Hmm. Yeah, we need the `--` in a lot of places to make sure the calls go through correctly. More than just here.

anomiex replied:
9527526

anomiex committed:
Add -- to pnpm run in scripts where passing args might be useful

anomiex committed:
doc: Canonicalize diff link

anomiex committed:
Turns out this override isn't necessary

Pnpm doesn't insist on upgrading it anymore now that some of that other
stuff is upgraded.

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

jeherve reviewed (commented):
Builds seem to work well for me. I just cannot figure out how to run Docker in the background anymore. With yarn, I used to do `yarn docker:up -d && yarn docker:jt-up` to start working. I wouldn't see the logs from Docker that way.

`pnpm docker:up -d` doesn't seem to work that way for some reason.

anomiex commented on .gitignore:
```diff
@@ -22,7 +22,6 @@ npm-debug.log
 .eslintcache
 .eslintignore-required
-yarn-error.log
 .phpcs.xml
```
Sure.
```suggestion
.phpcs.xml
.pnpm-debug.log
```

anomiex reviewed (commented):
It interprets the `-d` as a parameter to `pnpm` rather than passing it on to the script. You'd need to do `pnpm docker:up -- -d`. Or better, `jetpack docker up -d`.

anomiex committed:
Update .gitignore

zinigor commented:
Thanks for working on this update! I have been testing it in various scenarios, and I get a weird problem when trying to `npx jetpack install --all`:

```
❯ Installing plugins/backup
    ✔ Installing Composer Dependencies
    ✖ Installing Pnpm Dependencies
      →  ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-
…

  ----- SNIP -----

Error: Command failed with exit code 1: pnpm install
Scope: all 12 workspace projects
 ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-actions/required-review/node_modules/.bin/ncc'


    at makeError (/home/zinigor/workspace/jetpack/node_modules/.pnpm/execa@5.0.0/node_modules/execa/lib/error.js:59:11)
    at handlePromise (/home/zinigor/workspace/jetpack/node_modules/.pnpm/execa@5.0.0/node_modules/execa/index.js:114:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  shortMessage: 'Command failed with exit code 1: pnpm install',
  command: 'pnpm install',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: 'Scope: all 12 workspace projects\n' +
    " ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-actions/required-review/node_modules/.bin/ncc'\n" +
    '\n',
  stderr: '',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false,
  context: [Object: null prototype] {}
}
```

Strangely enough, it doesn't happen when I try to `npx jetpack install` and manually choose `projects/backup`.

jeherve commented:
> It interprets the `-d` as a parameter to `pnpm` rather than passing it on to the script. You'd need to do `pnpm docker:up -- -d`. Or better, `jetpack docker up -d`.

That worked, thank you!

For the Jurassic Tube part, I opened 2-gh-jurassic-tube

anomiex committed:
cli: Only run `pnpm install` once per run

Pnpm can race with itself if run multiple times in parallel for the same
workspace. Not terribly surprising, I guess.

anomiex commented:
@zinigor Looks like pnpm can race with itself if we run multiple installs for the same workspace at once. At least that's better than the situation with yarn, where it would race with itself even when running in different workspaces. a7a72b38 should avoid the multiple parallel runs.

P.S. You should probably use `pnpx` rather than `npx` there, but that doesn't seem to have anything to do with the problem you saw.

zinigor commented:
Thanks Brad! The last commit has fixed the installation for me!

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex committed:
Merge remote-tracking branch 'origin/master' into feature/replace-yarn-1-with-pnpm

anomiex commented:
r226560-wpcom
